service-base/src/main/java/org/eclipse/hono/service/plan/PrometheusBasedTenantPlanValidator.java

```diff
@@ -0,0 +1,125 @@
+     *
+     * @param vertx The Vertx instance to use.
+     * @throws NullPointerException if any of the parameters are {@code null}.
+     */
+    public PrometheusBasedTenantPlanValidator(final Vertx vertx) {
+        this.client = WebClient.create(Objects.requireNonNull(vertx));
+    }
+
+    @Override
+    public Future<?> isFurtherConnectionsAllowed(final TenantObject tenant) {
+        Objects.requireNonNull(tenant);
+        return checkConnectionsLimit(tenant);
+    }
+
+    /**
+     * Sets configuration properties.
+     *
+     * @param config The configuration properties.
+     * @throws NullPointerException if props is {@code null}.
+     */
+    @Autowired
+    public void setConfig(final PrometheusBasedTenantPlanValidatorProperties config) {
+        Objects.requireNonNull(config);
+        this.config = config;
+    }
+
+    private Future<?> checkConnectionsLimit(final TenantObject tenant) {
+        log.trace("Connections limit for tenant [{}] is [{}]", tenant.getTenantId(), tenant.getConnectionsLimit());
+        if (tenant.getConnectionsLimit() != -1) {
+            return queryForSumInPrometheus(MicrometerBasedMetrics.METER_CONNECTIONS_AUTHENTICATED.replace(".", "_"),
+                    tenant).recover(failure -> Future.succeededFuture(0L))
+                            .compose(noOfConnections -> {
+                                if (Optional.ofNullable(noOfConnections).orElse(0L) < tenant.getConnectionsLimit()) {
+                                    return Future.succeededFuture();
+                                } else {
+                                    log.trace(
+                                            "Connections limit exceeded for tenant. [tenant: {}, no.-Connections:{}, connections-limit:{}]",
+                                            tenant.getTenantId(), noOfConnections, tenant.getConnectionsLimit());
+                                    return Future.failedFuture(
+                                            String.format("Connections limit exceeded for tenant: [%s]",
+                                                    tenant.getTenantId()));
+                                }
+                            });
+        }
+        return Future.succeededFuture();
+    }
+
+    private Future<Long> queryForSumInPrometheus(final String metric, final TenantObject tenant) {
+        final Future<Long> result = Future.future();
+        final String query = String.format("sum(%s{tenant=\"%s\"})", metric, tenant.getTenantId());
+        log.debug("Prometheus backend url: {}, query: {}", config.getQueryURL(), query);
+        client.getAbs(config.getQueryURL())
+                .addQueryParam("query", query)
+                .send(sendResult -> {
+                    if (sendResult.succeeded()) {
+                        final HttpResponse<Buffer> response = sendResult.result();
+                        if (response.statusCode() == HttpResponseStatus.OK.code()) {
```

Comment:
you can make your life a little easier here by adding an expectation to the client on the result status
```
client.expect(ResponsePredicate.SC_OK)
```